### ISSUE TITLE
Support runtime deployment of non-annotated websocket endpoints

### ISSUE THREAD

**Issue by aaronanderson** [quarkusio/quarkus#3691]
**Describe the bug**
When attempting to register a web socket endpoint through an implementation of the ServerApplicationConfig interface an exception is generated.  The endpoint class implements the web socket Endpoint interface but it is not annotated with any web socket annotations.

**Expected behavior**
Quarkus/Undertow successfully deploys the websocket Endpoint implementation.

**Actual behavior**
The following exception is thrown:

> Caused by: java.lang.RuntimeException: javax.websocket.DeploymentException: UT003027: Class class graphql.servlet.GraphQLWebsocketServlet was not annotated with @ClientEndpoint or @ServerEndpoint
	at io.undertow.websockets.jsr.Bootstrap.handleDeployment(Bootstrap.java:93)
	at io.undertow.servlet.core.DeploymentManagerImpl.handleExtensions(DeploymentManagerImpl.java:279)
	at io.undertow.servlet.core.DeploymentManagerImpl.deploy(DeploymentManagerImpl.java:155)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder.bootServletContainer(UndertowDeploymentRecorder.java:468)
	... 12 more
Caused by: javax.websocket.DeploymentException: UT003027: Class class graphql.servlet.GraphQLWebsocketServlet was not annotated with @ClientEndpoint or @ServerEndpoint
	at io.undertow.websockets.jsr.ServerWebSocketContainer.addEndpointInternal(ServerWebSocketContainer.java:729)
	at io.undertow.websockets.jsr.ServerWebSocketContainer.addEndpoint(ServerWebSocketContainer.java:622)
	at io.undertow.websockets.jsr.Bootstrap.handleDeployment(Bootstrap.java:87)
	... 15 more

**To Reproduce**
1. Create a new Quarkus project with the following dependencies:
`
<dependencies>

		<dependency>
			<groupId>io.quarkus</groupId>
			<artifactId>quarkus-core</artifactId>
			<version>0.21.1</version>
		</dependency>
		
		  <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-arc</artifactId>
            <version>0.21.1</version>
        </dependency>
        
          <dependency>
            <groupId>javax.enterprise</groupId>
            <artifactId>cdi-api</artifactId>
              <version>2.0.SP1</version>
        </dependency>
		<dependency>
			<groupId>io.quarkus</groupId>
			<artifactId>quarkus-undertow</artifactId>
			<version>0.21.1</version>
		</dependency>
		
		<dependency>
			<groupId>io.quarkus</groupId>
			<artifactId>quarkus-undertow-websockets</artifactId>
			<version>0.21.1</version>
		</dependency>
		
		<dependency>
			<groupId>com.graphql-java-kickstart</groupId>
			<artifactId>graphql-java-servlet</artifactId>
			<version>8.0.0</version>
		</dependency>

	

	</dependencies>

`

2. Add the following class
```
import java.util.Set;

import javax.websocket.Endpoint;
import javax.websocket.HandshakeResponse;
import javax.websocket.server.HandshakeRequest;
import javax.websocket.server.ServerApplicationConfig;
import javax.websocket.server.ServerEndpointConfig;

import graphql.execution.AsyncExecutionStrategy;
import graphql.execution.SubscriptionExecutionStrategy;
import graphql.schema.GraphQLSchema;
import graphql.schema.StaticDataFetcher;
import graphql.schema.idl.RuntimeWiring;
import graphql.schema.idl.SchemaGenerator;
import graphql.schema.idl.SchemaParser;
import graphql.schema.idl.TypeDefinitionRegistry;
import graphql.servlet.GraphQLWebsocketServlet;
import graphql.servlet.config.DefaultExecutionStrategyProvider;
import graphql.servlet.config.DefaultGraphQLSchemaProvider;
import graphql.servlet.config.ExecutionStrategyProvider;
import graphql.servlet.core.GraphQLObjectMapper;
import graphql.servlet.core.GraphQLQueryInvoker;
import graphql.servlet.input.GraphQLInvocationInputFactory;

public class HelloServerApplicationConfig implements ServerApplicationConfig {

    @Override
    public Set<ServerEndpointConfig> getEndpointConfigs(Set<Class<? extends Endpoint>> endpointClasses) {
        String schema = "type Query{hello: String}";

        SchemaParser schemaParser = new SchemaParser();
        TypeDefinitionRegistry typeDefinitionRegistry = schemaParser.parse(schema);

        RuntimeWiring runtimeWiring = RuntimeWiring.newRuntimeWiring().type("Query", builder -> builder.dataFetcher("hello", new StaticDataFetcher("world!!"))).build();

        SchemaGenerator schemaGenerator = new SchemaGenerator();
        GraphQLSchema graphQLSchema = schemaGenerator.makeExecutableSchema(typeDefinitionRegistry, runtimeWiring);

        GraphQLInvocationInputFactory invocationInputFactory = GraphQLInvocationInputFactory.newBuilder(new DefaultGraphQLSchemaProvider(graphQLSchema)).build();

        ExecutionStrategyProvider executionStrategyProvider = new DefaultExecutionStrategyProvider(new AsyncExecutionStrategy(), new AsyncExecutionStrategy(), new SubscriptionExecutionStrategy());
        GraphQLQueryInvoker queryInvoker = GraphQLQueryInvoker.newBuilder().withExecutionStrategyProvider(executionStrategyProvider).build();

        GraphQLObjectMapper graphQLObjectMapper = GraphQLObjectMapper.newBuilder().build();//withGraphQLErrorHandler(new HelloGraphQLErrorHandler()).build();

        GraphQLWebsocketServlet servlet = new GraphQLWebsocketServlet(queryInvoker, invocationInputFactory, graphQLObjectMapper);
        ServerEndpointConfig endpointConfiguration = ServerEndpointConfig.Builder.create(GraphQLWebsocketServlet.class, "/graphql-ws").configurator(new MyWebSocketEndpointConfigurator(servlet)).build();

        return Set.of(endpointConfiguration);
    }

    @Override
    public Set<Class<?>> getAnnotatedEndpointClasses(Set<Class<?>> scanned) {
        return Set.of(GraphQLWebsocketServlet.class);
    }

    public static class MyWebSocketEndpointConfigurator extends ServerEndpointConfig.Configurator {
        private GraphQLWebsocketServlet servlet;

        public MyWebSocketEndpointConfigurator(GraphQLWebsocketServlet servlet) {
            this.servlet = servlet;
        }

        @Override
        public <T> T getEndpointInstance(Class<T> clazz) throws InstantiationException {
            return (T) servlet;
        }

        @Override
        public void modifyHandshake(ServerEndpointConfig sec, HandshakeRequest request, HandshakeResponse response) {
            servlet.modifyHandshake(sec, request, response);
        }

    }

}
```

3. Attempt to start Quarkus

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
 - Output of `java -version`: 
```openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
```
 - GraalVM version (if different from Java): N/A
 - Quarkus version or git rev: 0.21.1

**Additional context**
I would like to deploy the [GraphQLWebsocketServlet](https://github.com/graphql-java-kickstart/graphql-java-servlet/blob/master/src/main/java/graphql/servlet/GraphQLWebsocketServlet.java) web socket endpoint to Quarkus. It implements the websocket Endpoint interface and is not annotated with the ServerEndpoint annotation. Further this servlet should be gracefully stopped when the container is shutdown. I was able to create a [ServletContextListener](https://gist.github.com/aaronanderson/5a530150c25dbefad145d19a833146aa) but the web socket ServerContainer is not available as a servlet context attribute in Quarkus. I searched the Quarkus source code for a reference to the undertow io.undertow.websockets.jsr.ServerWebSocketContainer and Bootstrap classes related to the ServletContextEvent but I could not find any references.

**Comment by stuartwdouglas** [comment]
Can you remove the getAnnotatedEndpointClasses() method and see if this works?

This is not an annotated endpoint (but an Endpoint interface implementation) so it should not be returned from this method.

**Comment by aaronanderson** [comment]
Thanks for the tip! I returned an empty Set.of() in the getAnnotatedEndpointClasses() method and sure enough the endpoint was reported as being successfully deployed. 

[io.und.web.jsr] (main) UT026005: Adding programmatic server endpoint class graphql.servlet.GraphQLWebsocketServlet for path /graphql-ws

Also once Quarkus was aware of a websocket endpoint the "javax.websocket.server.ServerContainer" servlet context attribute was set and I was able to successfully register the endpoint a second time from the lifecycle listener. 

Since this servlet has a shutdown method that should be triggered I will experiment with combining the ServletContextListener and ServerApplicationConfig or create a placeholder endpoint so that the ServerContainer is always available in the lifecycle listener.
